### ISSUE TITLE
RMAction problem on swift3

### ISSUE THREAD

**Issue by illyaBakurov** [CooperRS/RMDateSelectionViewController#73]
I've updated the project to swift3 and was wondering when the support for this update is coming out?

I've converted the syntax with default xCode helper, but now when I add action to RMDateSelectionViewController xCode through error: 'RMAction' requires that 'RMActionController<UIDatePicker>' inherit from 'RMActionController<UIView>'

here is an example of ho wI use it:

if let cancelAction = RMAction(title: NSLocalizedString("Cancel", comment: "Button title on date picker selector"), style: .cancel, andHandler: nil) {
                dateSelectionController.addAction(cancelAction as! RMAction<RMActionController<UIDatePicker>>)
            }

**Comment by CooperRS** [comment]
Hi there,
thanks for reporting. I totally forgot about Swift 3 :D. I will take a lookt at the problem in the next days and come back to you when I found out something.
Roland

**Comment by rushafi** [comment]
Hi @CooperRS, thanks for all the good work on RMDateSelectionViewController. It was working great on my app, but after updating to Swift 3, I am getting the same error as @illyaBakurov. Any solution on this?

**Comment by CooperRS** [comment]
Hey,

I made a little SwiftDemo project which actually works for me with Swift 3 (https://github.com/CooperRS/RMDateSelectionViewController/tree/master/RMDateSelectionViewController-SwiftDemo). If you find any problems with using the demo project please let me know :)

Best regards,
Roland

**Comment by erusso1** [comment]
Hey cooper that swift 3 demo doesn't work for swift 3.

**Comment by CooperRS** [comment]
Can you give me the error that Xcode gives you? For me its working with Xcode 8.

**Comment by erusso1** [comment]
First make  sure you've migrated to current swift 3 syntax. Xcode 8 is still supporting older versions of swift.

The error is when creating a new instance of RMDateSelectionViewController, as in the following line: 

`guard let dateSelectionController = RMDateSelectionViewController(style: style, title: title, message: message, selectAction: selectAction, andCancelAction: cancelAction) else {return}`

This used to be fine on older versions of the compiler, but now we are all seeing this error:

"Cannot convert value of type 'RMAction<RMActionController<UIView>>? to expected argument of type 'RMAction<RMActionController<UIDatePicker>>''"

The arguments it is talking about are the select and cancel actions respectively, which were made using the following: 

`let selectAction = RMAction(title: "Select", style: .done) { controller in }`

`let cancelAction = RMAction(title: "Cancel", style: .cancel, andHandler: nil)`

**Comment by CooperRS** [comment]
Ah, I'm sorry. I updated the Swift demo project of RMActionController (the controller RMDateSelectionViewController relies on) to Swift 3(https://github.com/CooperRS/RMActionController/tree/master/RMActionController-SwiftDemo). You are right: I did not update the Swift demo of RMDateSelectionViewController yet.

**Comment by erusso1** [comment]
No worries! Glad to see the quick replies. I believe I found the underlying problem.

The error is that the compiler wants the actions to be RMAction < RMActionController < UIDatePicker > > ,  but instead we are giving it < RMActionController < UIView > > actions.

So I thought, let's try to cast them like this: 

`guard let s = selectAction as? RMAction<RMActionController<UIDatePicker>> else {return}`

`guard let c = cancelAction as? RMAction<RMActionController<UIDatePicker>> else {return}`

Now we see new errors, stating the following: 

"RMAction requires that RMActionController < UIDatePicker > inherit from RMActionController < UIView >"

This is strange, because UIDatePicker inherits from UIControl, which of course inherits from UIView. This could be part of the API changes to UIKit in Swift 3.

**Comment by deville** [comment]
I had the same problem and solved it by forking `RMDateSelectionViewController` and `RMActionController` and removing generics altogether. I don't really see why they are needed.

On order to use my forked version, add the following lines to your `Podfile`: 

```
  pod 'RMActionController',
    :git => 'git@github.com:deville/RMActionController.git',
    :branch => 'remove-generics'
  
  pod 'RMDateSelectionViewController',
    :git => 'git@github.com:deville/RMDateSelectionViewController.git',
    :branch => 'remove-generics'
```

**Comment by erusso1** [comment]
Thanks for the fork. Tried to update my pods but got a permission error: 

`[!] Error installing RMActionController
[!] /usr/bin/git clone git@github.com:deville/RMActionController.git /tmp/d20161002-6224-36pixt --template= --single-branch --depth 1 --branch remove-generics

Cloning into '/tmp/d20161002-6224-36pixt'...
Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.`

**Comment by deville** [comment]
@erusso1 strange, both repositories are public, branches are pushed. Try using HTTPS maybe.

**Comment by erusso1** [comment]
Thanks for the tip, did that and working perfect. Well done. @CooperRS Cooper should pull this in and make it live on a Swift3 branch.

@deville your solution is much appreciated.

**Comment by deville** [comment]
@erusso1 You're welcome :)
@CooperRS what do you think about generics? Are they really needed? If not, I will create pull request.

**Comment by jordanmaguire** [comment]
FWIW RMPickerViewController is afflicted by the same bug

**Comment by jlaws** [comment]
+1 on removing the generics (for now), the code is not usable in Swift 3 because Swift's type system is not yet able to support what you are trying to do.

**Comment by CooperRS** [comment]
Hey!

Sorry for the long delay, there is much to do :/.

I updated the master branch of both `RMActionController` and `RMDateSelectionViewController` and tweaked the generics a little bit be compatible with Swift 3. Unfortunately, this required a little API change. Basically, `RMAction` now only has a `UIView` subtype as generic instead of an `RMActionController` subtype. I'll write some documentation on this tomorrow :)

I'd prefere to fix the generics instead of removing them. Removing them means that casting is required in the callback blocks of `RMAction`. And casting is something I do not like :).

Best regards,
Roland

**Comment by CooperRS** [comment]
OK, version 2.1 of RMDateSelectionViewController is finished, available in the git repo (tag 2.1.0) and has been published to CocoaPods. This version should be compatible with Swift 3 😄

If you have any problem with this version please reopen this issue or create a new one.
